server/src/main/java/org/opensearch/search/approximate/ApproximatePointRangeQuery.java

```diff
@@ -248,14 +250,49 @@ private void intersectRight(PointValues.PointTree pointTree, PointValues.Interse
             // custom intersect visitor to walk the left of the tree
             public void intersectLeft(PointValues.IntersectVisitor visitor, PointValues.PointTree pointTree, long[] docCount)
                 throws IOException {
-                PointValues.Relation r = visitor.compare(pointTree.getMinPackedValue(), pointTree.getMaxPackedValue());
                 if (docCount[0] >= size) {
                     return;
                 }
+                // Outside query - quick exit
+                PointValues.Relation r = visitor.compare(pointTree.getMinPackedValue(), pointTree.getMaxPackedValue());
+                if (r == PointValues.Relation.CELL_OUTSIDE_QUERY) {
+                    return;
+                }
+                // Fast path for dense nodes
+                if (pointTree.size() > size && docCount[0] < size) {
+                    boolean processedDeepDive = false;
+                    // deep-dive
+                    if (pointTree.moveToChild()) {
+                        processedDeepDive = true;
+                        do {
+                            intersectLeft(visitor, pointTree, docCount);
+                            if (docCount[0] >= size) {
+                                pointTree.moveToParent();
+                                return;
+                            }
+                        } while (pointTree.size() > size && pointTree.moveToChild());
+                        pointTree.moveToParent();
+
+                        // If we've collected documents but not enough size, process the siblings
```

Comment:
Is this for the case where a single document has multiple values? So we may visit the left subtree based on the node size, but the actual number of doc IDs is small.

Maybe we could simplify by only applying this optimization when doc count >= point count?